modules/examples/sources/msg-generator/msg-generator-source.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2018 Balabit
+ * Copyright (c) 2018 László Várady <laszlo.varady@balabit.com>
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "msg-generator-source.h"
+
+#include "logsource.h"
+#include "logpipe.h"
+#include "timeutils.h"
+#include "seqnum.h"
+#include "logmsg/logmsg.h"
+#include "template/templates.h"
+
+#include <iv.h>
+
+struct MsgGeneratorSource
+{
+  LogSource super;
+  MsgGeneratorSourceOptions *options;
+  struct iv_timer timer;
+  gint32 seq_num;
+};
+
+static void
+_stop_timer(MsgGeneratorSource *self)
+{
+  if (iv_timer_registered(&self->timer))
+    iv_timer_unregister(&self->timer);
+}
+
+static void
+_start_timer(MsgGeneratorSource *self)
+{
+  iv_validate_now();
+  self->timer.expires = iv_now;
+  timespec_add_msec(&self->timer.expires, self->options->freq);
+
+  iv_timer_register(&self->timer);
+}
+
+static gboolean
+_init(LogPipe *s)
+{
+  MsgGeneratorSource *self = (MsgGeneratorSource *) s;
+  if (!log_source_init(s))
+    return FALSE;
+
+  _start_timer(self);
+
+  return TRUE;
+}
+
+static gboolean
+_deinit(LogPipe *s)
+{
+  MsgGeneratorSource *self = (MsgGeneratorSource *) s;
+
+  _stop_timer(self);
+
+  return log_source_deinit(s);
+}
+
+static void
+_free(LogPipe *s)
+{
+  log_source_free(s);
+}
+
+static void
+_send_generated_message(MsgGeneratorSource *self)
+{
+  if (!log_source_free_to_send(&self->super))
+    return;
+
+  LogMessage *msg = log_msg_new_empty();
+
```

Comment:
I would probably initialize this LogMessage to "-- Generated message. --" as it is initialized below and use that as an input for the template format operation.